src/main/java/org/jabref/logic/importer/fetcher/MergingIdBasedFetcher.java

```diff
@@ -0,0 +1,121 @@
+
+package org.jabref.logic.importer.fetcher;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.jabref.logic.importer.IdBasedFetcher;
+import org.jabref.logic.importer.ImportFormatPreferences;
+import org.jabref.logic.importer.fetcher.isbntobibtex.IsbnFetcher;
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.entry.field.Field;
+import org.jabref.model.entry.field.StandardField;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MergingIdBasedFetcher {
+    private static final List<StandardField> SUPPORTED_FIELDS =
+            List.of(StandardField.DOI, StandardField.ISBN, StandardField.EPRINT);
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MergingIdBasedFetcher.class);
+    private final ImportFormatPreferences importFormatPreferences;
+
+    public MergingIdBasedFetcher(ImportFormatPreferences importFormatPreferences) {
+        this.importFormatPreferences = importFormatPreferences;
+    }
+
+    public Optional<FetcherResult> fetchEntry(BibEntry entry) {
+        logEntryIdentifiers(entry);
+
+        return SUPPORTED_FIELDS.stream()
+                               .map(field -> tryFetch(entry, field))
+                               .filter(Optional::isPresent)
+                               .map(Optional::get)
+                               .findFirst();
+    }
+
+    private Optional<FetcherResult> tryFetch(BibEntry entry, Field field) {
+        return entry.getField(field)
+                    .flatMap(identifier -> getFetcherForField(field)
+                            .flatMap(fetcher -> fetchEntry(fetcher, field, identifier, entry)));
+    }
+
+    private Optional<FetcherResult> fetchEntry(IdBasedFetcher fetcher, Field field,
+                                               String identifier, BibEntry originalEntry) {
+        try {
+            LOGGER.debug("Attempting to fetch entry using {} fetcher", field);
+            logEntryDetails(originalEntry, field, identifier);
+
+            return fetcher.performSearchById(identifier)
+                          .map(fetchedEntry -> {
+                              BibEntry mergedEntry = (BibEntry) originalEntry.clone();
+                              boolean hasChanges = mergeBibEntry(mergedEntry, fetchedEntry);
+                              return new FetcherResult(originalEntry, mergedEntry, hasChanges);
+                          });
+        } catch (Exception exception) {
+            LOGGER.error("Error fetching entry with {} identifier: {}",
+                    field, identifier, exception);
+            return Optional.empty();
+        }
+    }
+
+    private boolean mergeBibEntry(BibEntry target, BibEntry source) {
```

Comment:
JavaDoc should say that all fields of target are overwritten by source value.

However, I would implement it differently: source is the fetcher bibentry, target is the existing bibentry., Overwrite only if target field is not existing.